### PR TITLE
[VL-105] - Endpoint to get payment notification document metadata

### PR DESCRIPTION
**Pull request by kin0992**
Add a new endpoint that returns the metadata of a payment notification document

#### How Has This Been Tested?

Unit tests

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/domain/GetPaymentNotificationMetadataRepository.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+export type GetPaymentNotificationMetadataRecord = {
+  type: 'GetPaymentNotificationMetadataRecord';
+  input: { apiKey: ApiKey; iun: Iun; recipientId: number; attachmentName: string };
+  output:
+    | Response<200, NotificationAttachmentDownloadMetadataResponse>
+    | Response<403, UnauthorizedMessageBody>
+    | Response<404>;
+};
+
+export type GetPaymentNotificationMetadataRepository = Repository<GetPaymentNotificationMetadataRecord>;
```
Comment (datalek) [inline src/domain/GetPaymentNotificationMetadataRepository.ts:16]:
```suggestion
export type GetPaymentNotificationMetadataRecordRepository = Repository<GetPaymentNotificationMetadataRecord>;
```

File: src/main.ts [hunk 2]
```diff
@@ -21,21 +21,26 @@ import { ConsumeEventStreamRecord } from './domain/ConsumeEventStreamRecordRepos
...
+    const uploadToS3RecordRepository = inMemory.makeRepository(logger)<UploadToS3Record>([]);
+    const newNotificationRepository = inMemory.makeRepository(logger)<NewNotificationRecord>([]);
+    const createEventStreamRecordRepository = inMemory.makeRepository(logger)<CreateEventStreamRecord>([]);
+    const checkNotificationStatusRepository = inMemory.makeRepository(logger)<CheckNotificationStatusRecord>([]);
+    const getNotificationDetailRepository = inMemory.makeRepository(logger)<GetNotificationDetailRecord>([]);
+    const consumeEventStreamRepository = inMemory.makeRepository(logger)<ConsumeEventStreamRecord>([]);
     const getNotificationDocumentMetadataRecordRepository = mkRepository<GetNotificationDocumentMetadataRecord>([]);
+    const getPaymentNotificationMetadataRepository = inMemory.makeRepository(
+      logger
+    )<GetPaymentNotificationMetadataRecord>([]);
```
Comment (datalek) [inline src/main.ts:43]: I've replaced `inMemory.makeRepository(logger)` with `mkRepository` which is a partially applied function

```suggestion
    const preLoadRecordRepository = mkRepository<PreLoadRecord>([]);
    const uploadToS3RecordRepository = mkRepository<UploadToS3Record>([]);
    const newNotificationRepository = mkRepository<NewNotificationRecord>([]);
    const createEventStreamRecordRepository = mkRepository<CreateEventStreamRecord>([]);
    const checkNotificationStatusRepository = mkRepository<CheckNotificationStatusRecord>([]);
    const getNotificationDetailRepository = mkRepository<GetNotificationDetailRecord>([]);
    const consumeEventStreamRepository = mkRepository<ConsumeEventStreamRecord>([]);
    const getNotificationDocumentMetadataRecordRepository = mkRepository<GetNotificationDocumentMetadataRecord>([]);
    const getPaymentNotificationMetadataRepository = mkRepository<GetPaymentNotificationMetadataRecord>([]);
```

Comment (kin0992) [reply src/main.ts:43]: My bad! I kept the wrong version during the merge! I'm going to fix it

File: src/domain/GetPaymentNotificationMetadataRepository.ts (hunk 1 above)
Comment (datalek) [reply src/domain/GetPaymentNotificationMetadataRepository.ts:16]: In my opinion, we should re-order all "Repository" items at some point (as a separate task)

Comment (kin0992) [reply src/domain/GetPaymentNotificationMetadataRepository.ts]: Done!

File: src/main.ts (hunk 2 above)
Comment (kin0992) [reply src/main.ts]: Changed it!

File: src/useCases/GetPaymentNotificationMetadataUseCase.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+          TE.ap(findNotificationRequestRecordRepository.list()),
+          TE.ap(consumeEventStreamRecordRepository.list()),
+          TE.map(
+            flow(
+              RA.filterMap(O.fromEither),
+              RA.chain((notification) => (notification.iun === iun ? notification.recipients : RA.empty)),
+              getRecipientPayment,
+              RA.map(getNotificationPaymentAttachment(attachmentName)),
+              RA.compact,
+              RA.last,
```
Comment (datalek) [inline src/useCases/GetPaymentNotificationMetadataUseCase.ts:76]:
```suggestion
              RA.findLastMap(getNotificationPaymentAttachment(attachmentName)),
```

File: openapi/index.yaml [hunk 4]
```diff
@@ -163,6 +163,36 @@ components:
...
+        internalId:
+          type: "string"
+          description: "id interno anonimizzato"
+        denomination:
+          type: "string"
+          description: "Nome e cognome / ragione sociale"
+        digitalDomicile:
+          $ref: "#/components/schemas/NotificationDigitalAddress"
+        physicalAddress:
+          $ref: "#/components/schemas/NotificationPhysicalAddress"
```
Comment (datalek) [inline openapi/index.yaml:193]:
```suggestion
        digitalDomicile:
          $ref: "https://raw.githubusercontent.com/pagopa/pn-delivery/0599f1287a08c315bc20410fc57a666e0fa6f79d/docs/openapi/schemas-pn-notification-v1.yaml#/components/schemas/NotificationDigitalAddress"
        physicalAddress:
          $ref: "https://raw.githubusercontent.com/pagopa/pn-delivery/0599f1287a08c315bc20410fc57a666e0fa6f79d/docs/openapi/schemas-pn-notification-v1.yaml#/components/schemas/NotificationPhysicalAddress"
```

File: openapi/index.yaml [hunk 5]
```diff
@@ -188,6 +218,99 @@ components:
...
+          description: "Comune"
+        municipalityDetails:
+          type: "string"
+          description: "Frazione o località"
+        province:
+          type: "string"
+          description: "Provincia"
+        foreignState:
+          type: "string"
+          description: "Denominazione paese estero"
```
Comment (datalek) [inline openapi/index.yaml:269]:
```suggestion
```

File: openapi/index.yaml [hunk 6]
```diff
@@ -188,6 +218,99 @@ components:
...
+      required:
+        - noticeCode
+        - creditorTaxId
+      properties:
+        noticeCode:
+          $ref: '#/components/schemas/noticeCode'
+        creditorTaxId:
+          $ref: '#/components/schemas/paTaxId'
+        noticeCodeAlternative:
+          $ref: '#/components/schemas/noticeCode'
```
Comment (datalek) [inline openapi/index.yaml:293]:
```suggestion
        noticeCode:
          $ref: "https://raw.githubusercontent.com/pagopa/pn-delivery/0599f1287a08c315bc20410fc57a666e0fa6f79d/docs/openapi/schemas-pn-notification-v1.yaml#/components/schemas/noticeCode"
        creditorTaxId:
          $ref: "https://raw.githubusercontent.com/pagopa/pn-delivery/0599f1287a08c315bc20410fc57a666e0fa6f79d/docs/openapi/schemas-pn-notification-v1.yaml#/components/schemas/paTaxId"
        noticeCodeAlternative:
          $ref: "https://raw.githubusercontent.com/pagopa/pn-delivery/0599f1287a08c315bc20410fc57a666e0fa6f79d/docs/openapi/schemas-pn-notification-v1.yaml#/components/schemas/noticeCode"          
```

File: openapi/index.yaml [hunk 7]
```diff
@@ -188,6 +218,99 @@ components:
...
+      maxLength: 18
+      minLength: 18
+      pattern: "^\\d+$"
+    paTaxId:
+      description: "Payment PA fiscal code"
+      example: "77777777777"
+      type: "string"
+      maxLength: 11
+      minLength: 11
+      pattern: "^\\d+$"
```
Comment (datalek) [inline openapi/index.yaml:313]:
```suggestion
```

**Review by datalek — CHANGES_REQUESTED** [review]
We can simplify the openapi referencing some resources in a different way

File: openapi/index.yaml (hunk 6 above)
Comment (datalek) [reply openapi/index.yaml:293]: Double check this suggestion, I didn't check it

File: openapi/index.yaml (hunk 7 above)
Comment (datalek) [inline openapi/index.yaml:313]: They must be defined in the following ways in order to be generated :/

```suggestion
    noticeCode:
      $ref: 'https://raw.githubusercontent.com/pagopa/pn-delivery/0599f1287a08c315bc20410fc57a666e0fa6f79d/docs/openapi/schemas-pn-notification-v1.yaml#/components/schemas/noticeCode'
    paTaxId:
      $ref: 'https://raw.githubusercontent.com/pagopa/pn-delivery/0599f1287a08c315bc20410fc57a666e0fa6f79d/docs/openapi/schemas-pn-notification-v1.yaml#/components/schemas/paTaxId'
```

File: openapi/index.yaml (hunk 6 above)
Comment (kin0992) [reply openapi/index.yaml:293]: In this case, accepting your suggestion, we will have a compile error.
The generator doesn't generate the proper type and issue an error on the import of those fields (e.g. noticeCode).

Comment (datalek) [reply openapi/index.yaml:244]: > In this case, accepting your suggestion, we will have a compile error. The generator doesn't generate the proper type and issue an error on the import of those fields (e.g. noticeCode).

Does [this](https://github.com/pagopa/pn-local-emulator/pull/50#discussion_r994674738) suggestion fix the issue?

Comment (kin0992) [reply openapi/index.yaml:244]: Yep, and it has been accepted

File: openapi/index.yaml (hunk 4 above)
Comment (kin0992) [reply openapi/index.yaml]: Nice! Changed

**Review by datalek — APPROVED** [review]
🥮

File: src/useCases/GetPaymentNotificationMetadataUseCase.ts [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        pipe(
+          TE.of(computeSnapshot(occurrencesAfterComplete, senderPAId, iunGenerator, dateGenerator)),
+          TE.ap(createNotificationRequestRecordRepository.list()),
+          TE.ap(findNotificationRequestRecordRepository.list()),
+          TE.ap(consumeEventStreamRecordRepository.list()),
+          TE.map(
+            flow(
+              RA.filterMap(O.fromEither),
+              RA.chain((notification) => (notification.iun === iun ? notification.recipients : RA.empty)),
+              getRecipientPayment,
```
Comment (datalek) [inline src/useCases/GetPaymentNotificationMetadataUseCase.ts:73]:
```suggestion
              RA.findLastMap(getNotificationPaymentAttachment(attachmentName)),
```

File: src/useCases/GetPaymentNotificationMetadataUseCase.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        return O.none;
+    }
+  };
+
+const getRecipientPayment = (recipients: Notification['recipients']): ReadonlyArray<NotificationPaymentInfo> =>
+  pipe(
+    recipients,
+    RA.map((recipient) => O.fromNullable(recipient.payment)),
+    RA.compact
+  );
```
Comment (datalek) [inline src/useCases/GetPaymentNotificationMetadataUseCase.ts:44]:
```suggestion
```

File: src/useCases/GetPaymentNotificationMetadataUseCase.ts [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+import {
+  GetPaymentNotificationMetadataRecord,
+  GetPaymentNotificationMetadataRecordRepository,
+  makePaymentNotificationAttachmentDownloadMetadataResponse,
+} from '../domain/GetPaymentNotificationMetadataRecordRepository';
+import { ApiKey } from '../generated/definitions/ApiKey';
+import { Iun } from '../generated/definitions/Iun';
+import { authorizeApiKey } from '../domain/authorize';
+import { computeSnapshot } from '../domain/Snapshot';
+import { Notification } from '../domain/Notification';
```
Comment (datalek) [inline src/useCases/GetPaymentNotificationMetadataUseCase.ts:19]:
```suggestion
```

**Review by datalek — APPROVED** [review]

**Pull request merged**